### ISSUE TITLE
Add logging messages to the password form protecting the dashboard to be used by Fail2ban

### ISSUE THREAD

**Issue by lamat1111** [SatoriNetwork/Neuron#15]
_I am not a professional dev, but this what Claude Sonnet suggest._

## Implement fail2ban-compatible logging for DDoS protection in Flask app

To enhance the security of our Flask application's login form against potential DDoS attacks, I suggest implementing fail2ban-compatible logging. This would allow us to use fail2ban on Linux systems to monitor and block suspicious IP addresses.

### Required Log Information

For effective fail2ban integration, our logs should include:

1. **IP Address**: The source IP of the request (crucial for fail2ban).
2. **Timestamp**: Helps in analyzing patterns and setting up advanced rules.
3. **Action/Event**: Clear indication of what happened (e.g., "Failed login attempt").
4. **Consistent Log Format**: Easily parseable by fail2ban's regex filters.

### Proposed Log Format

An ideal log entry format for fail2ban compatibility:

```
2024-07-29 14:30:15 - Failed login attempt | IP: 192.168.1.100
```

### Implementation Suggestions

1. **Logging Setup in Flask**:
   Add logging configuration to your Flask application:

   ```python
   import logging
   from logging.handlers import RotatingFileHandler
   import os

   log_dir = '/var/log/flask'
   if not os.path.exists(log_dir):
       os.makedirs(log_dir)
   log_file = os.path.join(log_dir, 'flask_auth.log')

   handler = RotatingFileHandler(log_file, maxBytes=10000, backupCount=1)
   handler.setLevel(logging.INFO)
   app.logger.addHandler(handler)
   ```

2. **Logging Failed Login Attempts**:
   In the route that handles the `VaultPassword` form submission, add logging for failed attempts:

   ```python
   from flask import request

   @app.route('/login', methods=['GET', 'POST'])
   def login():
       form = VaultPassword()
       if form.validate_on_submit():
           if check_password(form.password.data):  # Replace with your actual password check
               return redirect(url_for('dashboard'))
           else:
               app.logger.warning(f"Failed login attempt | IP: {request.remote_addr}")
               return render_template('login.html', form=form, error="Invalid password")
       return render_template('login.html', form=form)
   ```

3. **fail2ban Configuration**:
   After implementing logging, the user can, if they want, set up fail2ban with these configurations:

   Filter (`/etc/fail2ban/filter.d/flask-auth.conf`):
   ```
   [Definition]
   failregex = Failed login attempt \| IP: <HOST>
   ignoreregex =
   ```

   Jail (`/etc/fail2ban/jail.local`):
   ```
   [flask-auth]
   enabled = true
   port = http,https
   filter = flask-auth
   logpath = /var/log/flask/flask_auth.log
   maxretry = 5
   bantime = 3600
   ```

**Comment by Tylerlhess** [comment]
added this functionality in https://github.com/SatoriNetwork/Neuron/pull/16 the config variable that needs to be set is "fail2ban_log_dir". There currently are no configuration instructions for the fail2ban side but the instructions provided above seem adequate.

**Comment by lastmeta** [comment]
@lamat1111 
Do we want logs about someone attempting to gain access or do we want to punish them for trying to gain access? what if we added a back off period after every attempt? the real owner could get locked out from repeated attempts, but they have access to the os through ssh so they can restart the flask app and get in if they have to. thoughts?

**Comment by lamat1111** [comment]
Well, my idea was just to add logging messages for the Satori dashboard form, and then the user can set fail2ban as they want by themselves.
For instance: monitor for repeated login attempts failed, if more than 3 in 5 minutes, then lock out IP for 1 hour.
If then you want to somewhat code the fail2ban settings in Satori, ok, but I think any user needing this should be able to take care of fail2ban themselves, no?

**Comment by Tylerlhess** [comment]
yeah fail2ban can be set to the port specifically. So they get blocked to the webui port. I agree with lamat1111. That all thats needed is the log. It probably could have even just been a log line in the general logs. In that case this is way overkill but I like having the access logs separate.

**Comment by lastmeta** [comment]
> I think any user needing this should be able to take care of fail2ban themselves, no?

I think if we can default some defense solution for them we should. the timeout covers the situation of getting attacked from multiple ips. the only trade off is the attacker can time out the user so the user has to restart the neuron to get immediate access, an inconvenience, but I think a smaller one than getting brute forced and a smaller one than setting up my own fail2ban rules, even if I know what I'm doing its an extra step everyone would have to take to gain the security they should have already.

is there a downside to this one-size-fits-all solution I'm missing?

**Comment by Tylerlhess** [comment]
No theres not really a downside. We can put fail2ban directly in the docker container and have it preconfigured. It works great and does what we are wanting it to do. If the user wants to turn it off they can through configs. All we have to do is make sure we are setting it up and its working. The only disadvantage of putting it in the docker container is it won't also protect ssh ports and would need to be setup there also if they wanted a full security package.

**Comment by lamat1111** [comment]
> The only trade off is the attacker can time out the user so the user has to restart the neuron to get immediate access, an inconvenience, but I think a smaller one than getting brute forced and a smaller one than setting up their own fail2ban rules, (which they could still do I guess), even if they know what they're doing its an extra step everyone would have to take to gain the security they should have already.

So with this solution the login access will be blocked when a N number of login fails from any IP (not just the same one)?
I wonder if this could become an issue if one gets heavily attacked?

Another security step, if not too complex, would be to give the user the possibility to set a custom port number to access the dashboard, or set a random one by default maybe?

**Comment by lastmeta** [comment]
I added the fail to ban logs, without any pre-configuration, so those that want to can use it. And I added a default backoff maximum period of 5 minutes. giving all attackers a communal number of 288 guesses on someone's neuron lock password a day. Not enough to brute force it, if it's at all a secure password.

Thanks for bringing this issue up.
